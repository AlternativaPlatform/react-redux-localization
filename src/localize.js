import { createSelector } from 'reselect';
import { connect } from 'react-redux';
import defaultTranslator from './default-translator';
import { getLocale } from './locale-reducer';

export default (
  translations,
  mapStateToLocale = getLocale,
  propName = 'l',
  translator = defaultTranslator,
) => {
  const makeTranslator = createSelector(
    locale => locale,
    locale => key => translator(translations, key, locale)
  );

  const mapStateToProps = state => ({
    [propName]: makeTranslator(mapStateToLocale(state)),
  });

  return WrappedComponent => connect(
    mapStateToProps,
    null,
    null,
    { getDisplayName: name => `Localized(${name})` }
  )(WrappedComponent);
}

