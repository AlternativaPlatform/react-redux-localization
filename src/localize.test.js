import React from 'react';
import testRenderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import localize from './localize';

const mockStore = configureStore();
const defaultTranslations = { hello: { de: 'Hallo, {}!' } };
const customTranslations = { de: { hello: 'Hallo, {}!' } };

console.log(Provider);
const render = (Component, storeState) => testRenderer.create(
  <Provider store={mockStore(storeState)}>
    <Component />
  </Provider>
);

test('localize with default settings', () => {
  const rendered = render(
    localize(defaultTranslations)(
      ({ l }) => <div>{l('hello', 'Fedor')}</div>
    ),
    { locale: 'de' }
  );
  expect(rendered.toJSON()).toMatchSnapshot();
});

test('localize with overriden settings', () => {
  console.log(customTranslations);
  const rendered = render(
    localize(
      customTranslations,
      state => state.language,
      'translate',
      (translations, key, locale, ...args) => translations[locale][key].replace('{}', args[0])
    )(
      ({ translate }) => <div>{translate('hello', 'Fedor')}</div>
    ),
    { language: 'de' }
  );
  expect(rendered.toJSON()).toMatchSnapshot();
});
