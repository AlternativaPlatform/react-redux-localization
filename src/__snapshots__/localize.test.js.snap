// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`localize with default settings 1`] = `
<div>
  Hallo, Fedor!
</div>
`;

exports[`localize with overriden settings 1`] = `
<div>
  Hallo, Fedor!
</div>
`;
